components/sup/src/manager/user_config_watcher.rs

```diff
@@ -0,0 +1,338 @@
+pub trait Serviceable {
+    fn name(&self) -> &str;
+    fn path(&self) -> &Path;
+    fn service_group(&self) -> &ServiceGroup;
+}
+
+impl Serviceable for Service {
+    fn name(&self) -> &str {
+        &self.pkg.name
+    }
+
+    fn path(&self) -> &Path {
+        &self.pkg.svc_path
+    }
+
+    fn service_group(&self) -> &ServiceGroup {
+        &self.service_group
+    }
+}
+
+
+// WorkerState contains the channels the worker uses to communicate with the Watcher.
+struct WorkerState {
+    // This receiver is used by the watcher to be notified when a worker has events.
+    // The channel is a SyncChannel with buffer size 1, as we are only interested in the fact that there were events,
+    // not how many there were.
+    have_events: Receiver<()>,
+    // This sender is used by the watcher to notify a worker to stop running.
+    // It is an async channel because we never want the UserConfigWatcher to block, even if the receiver end of
+    // the channel somehow dies and/or fails to consume the message.
+    stop_running: Sender<()>,
+}
+
+type ServiceName = String;
+pub struct UserConfigWatcher {
+    states: HashMap<ServiceName, WorkerState>,
+}
+
+impl UserConfigWatcher {
+    pub fn new() -> Self {
+        Self { states: HashMap::new() }
+    }
+
+    /// Adds a service to the User Config Watcher, thereby starting a watcher thread.
+    pub fn add<T: Serviceable>(&mut self, service: &T) -> io::Result<()> {
+        // It isn't possible to use the `or_insert_with` function here because it can't have a
+        // return value, which we need to return the error from `Worker::run`.
+        if let None = self.states.get(service.name()) {
+            // Establish bi-directional communication with the worker by creating two channels.
+            // The sync_channel's buffer size is 1 because we want to use it as a boolean, i.e. we
+            // are not interested in the events themselves, but only whether at least one has
+            // happened.
+            let (events_tx, events_rx) = sync_channel(1);
+            let (running_tx, running_rx) = channel();
+
+            Worker::run(&service.path(), events_tx, running_rx)?;
+
+            outputln!(preamble service.service_group(), "Watching {}", USER_CONFIG_FILE);
+
+            let state = WorkerState {
+                have_events: events_rx,
+                stop_running: running_tx,
+            };
+
+            self.states.insert(service.name().to_owned(), state);
+        }
+
+        Ok(())
+    }
+
+    /// Removes a service from the User Config Watcher, thereby stopping the watcher thread.
+    pub fn remove<T: Serviceable>(&mut self, service: &T) -> Result<(), SendError<()>> {
+        if let Some(state) = self.states.remove(service.name()) {
+            state.stop_running.send(())?;
+        }
+
+        Ok(())
+    }
+
+    /// Checks whether the watcher for the specified service has observed any events, thereby
```

Comment:
thereby is your new favourite word, eh? :)

Reply:
It's a pretty cool word, you gotta admit that